File: apps/app/lib/helplines.ts
/* Crisis helplines — served by the ENGINE, never hardcoded here.
 *
 * ARCHITECTURE.md §Cross-stack contracts: "never hardcoded in clients". The Zen
 * reference breaks this in exactly this file's place — its web chat ships India's KIRAN
 * number as a client-side fallback (`chat/page.tsx:154`) — and our own Android app shipped
 * the same bug until 2026-07-16: one country's numbers shown to everyone, next to a region
 * picker whose answer nothing read. Do not reintroduce it here.
 *
 * NEUTRAL is the only list in this file, it names NO country, and it is byte-identical to
 * what the engine returns for an unknown region (`app/safety/helplines.py::_INTERNATIONAL`).
 * It exists so the panel renders something dialable even if the fetch fails — a crisis
 * surface with nothing on it is the worst outcome available. Mirrors
 * `apps/android/.../data/Helplines.kt`; the two are deliberately the same shape.
 */

import { accessToken } from "./api";

const ENGINE = process.env.NEXT_PUBLIC_ENGINE_URL || "http://localhost:8000";

export type Helpline = {
  name: string;
  detail: string;
  /** A phone number to dial or a URL to open — `kind` says which, so we never sniff it. */
  target: string;
  kind: "tel" | "url";
};

/** The floor. Region-neutral by construction — see the file header. */
export const NEUTRAL: Helpline[] = [
  {
    name: "Find a helpline in your country",
    detail: "International directory · routes to your region",
    target: "https://findahelpline.com",
    kind: "url",
  },
];

/** Drop malformed rows rather than render a row that does nothing when tapped. */
export function parse(rows: unknown): Helpline[] {
  if (!Array.isArray(rows)) return [];
  return rows.flatMap((r) => {
    const o = r as Partial<Helpline> | null;
    if (!o || typeof o.target !== "string" || !o.target.trim()) return [];
    if (o.kind !== "tel" && o.kind !== "url") return [];
    return [{ name: String(o.name ?? ""), detail: String(o.detail ?? ""), target: o.target, kind: o.kind }];
  });
}

/**
 * The helplines to show. Never throws and never returns an empty list: every failure
 * path lands on NEUTRAL.
 *
 * `region` is the platform's resolved `crisis_region` from /users/me (the person's own
 * choice, else their org's default, else "" — which the engine answers with the
 * international directory rather than a guess).
 */
export async function loadHelplines(region: string): Promise<Helpline[]> {
  try {
    const token = await accessToken();
    // Unauthenticated is not a reason to show nothing — fall through to NEUTRAL.
    if (!token) return NEUTRAL;
    const r = await fetch(`${ENGINE}/v1/safety/helplines?region=${encodeURIComponent(region)}`, {
      headers: { Authorization: `Bearer ${token}` },
    });
    if (!r.ok) return NEUTRAL;
    const parsed = parse((await r.json())?.helplines);
    return parsed.length ? parsed : NEUTRAL;
  } catch {
    return NEUTRAL;
  }
}

/** What an <a> should point at. `tel:` for a number, the URL itself for a link. */
export const hrefFor = (h: Helpline) => (h.kind === "tel" ? `tel:${h.target.replace(/[^0-9+]/g, "")}` : h.target);

File: apps/android/app/src/main/java/com/cerebrozen/app/data/Helplines.kt
package com.cerebrozen.app.data

/* Crisis helplines — served by the engine, never hardcoded here.
 *
 * ARCHITECTURE.md §Cross-stack contracts already required this ("Crisis regions/helplines
 * config | Engine | Android crisis screen | never hardcoded in clients"). The app violated
 * it: one country's numbers were literals in Crisis.kt / Settings.kt / strings.xml, while
 * Settings offered a region picker whose answer nothing read. A user in the UK could pick
 * "GB" and still be shown India's Tele-MANAS. That is the bug this file closes.
 *
 * OFFLINE IS THE HARD PART, and it is why the numbers were inlined in the first place — a
 * crisis screen that needs the network is worse than one with wrong numbers. So there are
 * three layers, and the screen always renders:
 *
 *   1. the engine's answer for the person's region (Session caches every GET);
 *   2. the last cached answer, served automatically by Session on a connectivity failure;
 *   3. NEUTRAL below — the built-in floor for a first run with no network.
 *
 * NEUTRAL is deliberately identical to what the engine returns for an unknown region
 * (app/safety/helplines.py::_INTERNATIONAL): an international finder that routes to the
 * caller's own country. It names NO country, because on this path we do not know one —
 * and a confident guess is exactly the failure being fixed. Keep it that way: any
 * country's number added here is a regression, and HelplinesTest fails on it.
 */

import com.cerebrozen.app.net.Api
import org.json.JSONArray
import org.json.JSONObject

/** One dialable row. [kind] is "tel" or "url" — the engine says which, so this client
 *  never sniffs a target to decide how to act on it. */
data class Helpline(
    val name: String,
    val detail: String,
    val target: String,
    val kind: String,
) {
    val isUrl: Boolean get() = kind == "url"
}

object Helplines {
    /** The floor. Region-neutral by construction — see the file header. */
    val NEUTRAL: List<Helpline> = listOf(
        Helpline(
            name = "Find a helpline in your country",
            detail = "International directory · routes to your region",
            target = "https://findahelpline.com",
            kind = "url",
        ),
    )

    internal fun parse(rows: JSONArray): List<Helpline> =
        (0 until rows.length()).mapNotNull { i ->
            val o: JSONObject = rows.optJSONObject(i) ?: return@mapNotNull null
            val target = o.optString("target")
            val kind = o.optString("kind")
            // Drop anything malformed rather than render a row that does nothing when
            // tapped. If that empties the list, load() falls back to NEUTRAL.
            if (target.isBlank() || (kind != "tel" && kind != "url")) return@mapNotNull null
            Helpline(o.optString("name"), o.optString("detail"), target, kind)
        }

    /**
     * The helplines to show. Never throws and never returns an empty list: on this screen
     * an exception or a blank list is the worst available outcome, so every failure path
     * lands on [NEUTRAL].
     *
     * [region] comes from the platform's resolved `crisis_region` on /users/me (the
     * person's own choice, else their org's default, else "" for unknown).
     */
    suspend fun load(region: String): List<Helpline> {
        val parsed = runCatching {
            val body = JSONObject(Api.helplinesRaw(region))
            parse(body.optJSONArray("helplines") ?: JSONArray())
        }.getOrDefault(emptyList())
        return parsed.ifEmpty { NEUTRAL }
    }
}
